new/node_state_separation.md

```diff
@@ -0,0 +1,350 @@
+
+##### `GET /nodes/:name`
+
+This method will return the combined desired and current data and its semantics
+will be unchanged.
+
+##### `PUT /nodes/:name`
+
+This method will create a new node with the combined desired and current data
+and its semantics will be unchanged.
+
+##### `DELETE /nodes/:name`
+
+This method will delete a node and its semantics will be unchanged.
+
+#### `/nodes`
+
+This endpoint is unchanged, and is only included here for context.
+
+##### `POST /nodes`
+
+This method will create a new node object and its semantics will be unchanged.
+
+##### `GET /nodes`
+
+This method returns a list of all node URIs on the server and its semantics
+will be unchanged.
+
+### Core Chef Server Changes
+
+#### Nodes As Composed Objects
+
+The implementation of this will break up the node object into different objects
+for the current and desired state.  This will imply a migration to the new code
+  which will take existing node objects in a server and will break them up into
+  the new current and desired objects in the database.
+
+The ACLs on the node will also be composed of the ACLs on the underlying
+current and desired objects.  As part of the data migration on the server
+upgrade the ACLs on the node objects will be copied to both the desired and the
+current state (preserving the default behavior that nodes may update their own
+desired state).  The ACLs on the composed endpoints (e.g. `GET /nodes/:name`)
+will be composed of the ACLs on the underlying desired and current state and
+will require both of those to be allowed for the operation to succeed.  There
+will be no support for getting partial results back if the client has only
+partial permissions to the node components.
+
+### Core Chef Client Changes
+
+#### Node Retrieval
+
+The initial request from the chef-client will be changed to GET the
+`nodes/:name/desired` endpoint to retrieve only the saved desired data.  This
+avoids having the client retrieve all the node data, parse the JSON, inflate
+the attributes and then remove the normal, override and automatic attributes --
+saving time, bandwidth and memory.
+
+A copy of the desired data will be saved so that it can later be determined if
+the data has been "dirtied" during the chef-client run.
+
+For backwards compatibility if the prior request to the desired endpoint 404s,
+then the client will retrieve its node data from the old `/nodes/:name`
+endpoint and will fallback to backwards compatibility mode for all further
+operations.
+
+If the node does not exist it will still be created with a POST to `/nodes`
+
+#### Node Save
+
+Provided the desired endpoint is supported, the Chef::Node#save method will be
+updated to typically PUT its data only to the `/nodes/:name/current` endpoint.
+This will affect both the final `node.save` of the chef-client run, and any
+user called `node.save` in recipe code.
+
+If it is detected that the desired data has been dirtied then `node.save` will
```

Comment:
what does this mean? How does one detect dirty? Since the client gets a composed object, how does it no which pieces are from current vs desired?

Finally... when I request a runlist at the beginning of a run using the client, at the moment I don't believe that actually affects the node, it means the server does the dependency resolution based on that data instead of what it knows about the client... in the new model does that still work?